pallets/gear/src/internal.rs

```diff
@@ -317,28 +333,32 @@ where
     }
 
     /// Adds dispatch into waitlist, deposits event and adds task for waking it.
-    pub(crate) fn wait_dispatch(dispatch: StoredDispatch, reason: MessageWaitedReason) {
-        // Figuring out maximal deadline of holding.
-        let hold =
-            HoldBound::<T>::by(CostsPerBlockOf::<T>::waitlist()).maximum_for_message(dispatch.id());
+    pub(crate) fn wait_dispatch(
+        dispatch: StoredDispatch,
+        duration: Option<BlockNumberFor<T>>,
+        reason: MessageWaitedReason,
+    ) {
+        // `HoldBound` cost builder.
+        let hold_builder = HoldBound::<T>::by(CostsPerBlockOf::<T>::waitlist());
+
+        // Maximal hold bound for the message.
+        let maximal_hold = hold_builder.clone().maximum_for_message(dispatch.id());
+
+        // Figuring out correct hold bound.
+        let hold = if let Some(duration) = duration {
+            hold_builder.duration(duration).min(maximal_hold)
```

Comment:
It should be `duration` for `wait_for` call, shouldn't it?

Reply:
Not actually.
`Wait` result of execution contains `Option<BlockNumber>`. If it equals `None`, than we wait as much as possible (`maximum_for_message`), otherwise we wait minimum from as much as possible and maximal queried duration. In general, this is more about `wait_no_more`, but suitable for `wait_for` also, because it will panic in case of not enough funds. So I decided to keep this "check" for `wait_for` also just to make sure that there is enough funds one more time. (It can be replaced with some enum with appropriate to wait types variants, but it's just code repetition for me).